### PR TITLE
set tcp keepalive for tcp socket

### PR DESCRIPTION
**Pull request by juntaoliu**

### CODE REVIEW

**Comment by winlinvip** [comment]
Fix #1044 Issue of abnormal disconnection of TCP connection in live streaming pull flow end cannot be released.

`TRANS_BY_GPT3`

File: trunk/src/app/srs_app_listener.cpp [hunk 1]
```diff
@@ -223,6 +223,16 @@ int SrsTcpListener::listen()
     }
     srs_verbose("setsockopt reuse-addr success. port=%d, fd=%d", port, _fd);
     
+#ifdef SO_KEEPALIVE
```
Comment (winlinvip) [inline trunk/src/app/srs_app_listener.cpp:226]: Why use macro definition to enable this setting? Why not always set SO_KEEPALIVE?
This macro should be undefined, so this code will not take effect, right?

`TRANS_BY_GPT3`

Comment (winlinvip) [reply trunk/src/app/srs_app_listener.cpp]: 我改成了SRS_PERF_SO_KEEPALIVE，默认不开启。SRS2就这样了。

Comment (winlinvip) [reply trunk/src/app/srs_app_listener.cpp:226]: This macro is system-defined and is a reasonable way to write it.

`TRANS_BY_GPT3`

Comment (juntaoliu) [reply trunk/src/app/srs_app_listener.cpp:226]: This is not set to open using macros, but to enhance the portability of the code. Most Unix-like systems generally define SO_KEEPALIVE in the system, so there is no problem with or without this macro judgment in such systems. However, if the code is ported to other systems, those systems may not support SO_KEEPALIVE.

`TRANS_BY_GPT3`

**Pull request merged**